demo.ipynb

```diff
@@ -0,0 +1,1193 @@
+    "\n",
+    "The concept of a privacy budget is important in DP. Interactively developing DP analyses is a good practice, for example, to identify variables of interest before choosing where to spend the rest of a finite privacy budget. That said, it is important to track the total budget consumed: It would not be good to run a cell in this notebook, look at output, and then run it again, without tracking the privacy consumed by each operaion.\n",
+    "\n",
+    "We begin with a conventional set of imports:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 6,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Uncomment if you have not already installed OpenDP:\n",
+    "# %pip install opendp==0.11.1\n",
+    "\n",
+    "# OpenDP extends the Polars dataframe library:\n",
+    "import polars as pl\n",
+    "# Conventional import for OpenDP:\n",
+    "import opendp.prelude as dp\n",
+    "\n",
+    "# OpenDP plans to offer a set algorithms that have been formally reviewed for correctness.\n",
+    "# Until that is complete, we need to opt-in to \"contributed\" features:\n",
+    "dp.enable_features(\"contrib\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Next, the information that was provided in the UI:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 7,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Information you've provided that applies to the whole calculation:\n",
+    "csv_path = '/tmp/demo.csv'\n",
+    "contributions = 10\n",
+    "epsilon = 2\n",
+    "weights = [4, 4, 1, 1]\n",
+    "max_possible_rows = 1000000 \n",
+    "delta = 1e-7\n",
+    "\n",
+    "# Information you've provided for the \"grade\" column:\n",
+    "grade_min = 50\n",
+    "grade_max = 100\n",
+    "grade_bins_count = 10\n",
+    "\n",
+    "grade_bins_list = list(range(\n",
+    "    grade_min,\n",
+    "    grade_max,\n",
+    "    int((grade_max - grade_min + 1) / grade_bins_count)\n",
+    "))\n",
+    "\n",
+    "# Information you've provided for the \"class_year\" column:\n",
+    "class_year_min = 1\n",
+    "class_year_max = 4\n",
+    "class_year_bins_count = 4\n",
+    "\n",
+    "class_year_bins_list = list(range(\n",
+    "    class_year_min,\n",
+    "    class_year_max,\n",
+    "    int((class_year_max - class_year_min + 1) / class_year_bins_count)\n",
+    "))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Finally, define a [`Context`](https://docs.opendp.org/en/stable/api/user-guide/context/index.html) using the provided information."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 8,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "context = dp.Context.compositor(\n",
+    "    data=pl.scan_csv(csv_path, encoding=\"utf8-lossy\").with_columns(\n",
+    "        # The cut() method returns a Polars categorical type.\n",
```

Comment:
> I assume this is living outside because we haven't cut a nightly yet?

For this project, I'd prefer not to use nightlies. This is supposed to model what end users should do, and I think end users should be pinning versions for stability. (If there's a better way to do this with the next release, great!)

Reply:
I remember this comment now. You'd literally need to specify dummy bins like "A", "B", "C", and so on, which isn't as compelling as numeric ranges.